### PR TITLE
Use pagination for '/sponge plugins' command

### LINKED ISSUE #1239: Sponge Plugins list String too big
running `/sponge plugins` on big packs can kick you from the server

`[16:01:13] [Netty Server IO #1/ERROR]: io.netty.handler.codec.EncoderException: String too big (was 36841 bytes encoded, max 32767)`

old issue https://github.com/SpongePowered/SpongeForge/issues/918

### PR DESCRIPTION
**Pull request by Grinch**
**SpongeCommon** | **[SpongeForge](https://github.com/SpongePowered/SpongeForge/pull/1719)**

Fixes #1239, Supersedes #1250.

Base work done by @zachoooo.

**Screenshots**

### CODE REVIEW

**Comment by parlough** [comment]
I don't really see a need for the numbers and the light blue doesn't seem to fit that well. Maybe try light green instead?

**Comment by Grinch** [comment]
@Meronat Is [this](http://i.imgur.com/KTWJtBf.png) a bit better?

**Comment by vectrixdevelops** [comment]
@Grinch if I may make a suggestion, maybe when you hover over a plugin, you could also add a line for that plugins dependencies and the ones that aren't optional and are not present could show up red to assist the user in understanding what is missing. Not sure. 😄

**Comment by Grinch** [comment]
@connorhartley Pretty sure when a plugin loads, and it is missing a non-optional dependency that it will throw an exception and not load further.

**Comment by vectrixdevelops** [comment]
@Grinch true, but on the current list the plugin shows up as red to say it couldn't load. Not sure if this is still the effect you have on this list. If not, I wouldn't worry about my suggestion. :P

**Comment by Grinch** [comment]
@connorhartley This does not take place in the current code of SpongeCommon: https://github.com/SpongePowered/SpongeCommon/blob/bleeding/src/main/java/org/spongepowered/common/command/SpongeCommand.java#L500

**Edit**: Make no mistake, I like your idea. I'm just not sure how one even checks if a plugin loaded successfully since it wouldn't be registered I imagine as well as the tooltips in Minecraft are finicky as it is so adding a list of dependencies could result in some plugins having really bad tooltips.

**Comment by vectrixdevelops** [comment]
@Grinch Yeah, I am not sure why I am thinking it does that when there is absolutely no way that could happen. Oh well.

**Edit:**  I have bamboozled myself.

File: src/main/java/org/spongepowered/common/SpongeImplHooks.java [hunk 1]
```diff
@@ -334,4 +345,31 @@ public static void onCraftingRecipeRegister(CraftingRecipe recipe) {
...
+
+    public static Text getAdditionalCommandDescriptions() {
+        return Text.EMPTY;
+    }
+
+    public static void registerAdditionalCommands(ChildCommandElementExecutor flagChildren, ChildCommandElementExecutor nonFlagChildren) {
+        // Overwritten in SpongeForge
+    }
+
+    public static List<PluginContainer> getSortedPluginList() {
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/SpongeImplHooks.java]: Why is this here? It's only used in the command.

File: src/main/java/org/spongepowered/common/command/SpongeCommand.java [hunk 2]
```diff
@@ -472,64 +481,88 @@ private static CommandSpec getPluginsCommand() {
...
-                        }
-                    } else {
-                        Collection<PluginContainer> plugins = SpongeImpl.getGame().getPluginManager().getPlugins();
-                        Text.Builder build = Text.builder(String.format("Plugins (%d): ", plugins.size()));
-                        boolean first = true;
-                        for (PluginContainer next : plugins) {
-                            if (!first) {
-                                build.append(SEPARATOR_TEXT);
+                            final PaginationList.Builder builder = PaginationList.builder();
+                            builder.title(Text.of(TextColors.YELLOW, "Plugins", TextColors.WHITE, " (", sortedContainers.size(), ")"))
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/command/SpongeCommand.java]: Perhaps `GOLD` instead?

File: src/main/java/org/spongepowered/common/SpongeImplHooks.java [hunk 3]
```diff
@@ -334,4 +345,31 @@ public static void onCraftingRecipeRegister(CraftingRecipe recipe) {
...
+        final List<String> pluginStatics = Lists.newArrayList("minecraft", "mcp", "spongeapi", "sponge");
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/SpongeImplHooks.java]: Can be a constant.

File: src/main/java/org/spongepowered/common/SpongeImplHooks.java (hunk 1 above)
Comment (Grinch) [reply src/main/java/org/spongepowered/common/SpongeImplHooks.java]: It is overridden in SpongeForge to change the static entries at the beginning of the list.

File: src/main/java/org/spongepowered/common/command/SpongeCommand.java (hunk 2 above)
Comment (Grinch) [reply src/main/java/org/spongepowered/common/command/SpongeCommand.java]: Zidane wanted whatever is generally used for Sponge elsewhere.

File: src/main/java/org/spongepowered/common/SpongeImplHooks.java (hunk 1 above)
Comment (Grinch) [reply src/main/java/org/spongepowered/common/SpongeImplHooks.java]: To clarify actually. It is overwritten by SpongeForge to remove any mods from the list.

**Pull request merged**